### PR TITLE
Readbility edits: JavaScript Language section

### PR DESCRIPTION
**Pull request by alice**
***
<a href="https://pr-preview.s3.amazonaws.com/alice/design-principles/pull/222.html" title="Last updated on Sep 22, 2020, 8:12 AM UTC (95ef53e)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3ctag/design-principles/222/d0ad23a...alice:95ef53e.html" title="Last updated on Sep 22, 2020, 8:12 AM UTC (95ef53e)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+<div class="note">
+JavaScript is also known as [[!ECMASCRIPT]]
+</div>
 
-As such, when designing your APIs, your primary concern should be with the interface you present to
-JavaScript developers. You can freely rely upon language-specific semantics and conventions, with
-no need to keep things generalized.
+<div class="note">
+[[WEBIDL]] defines a separate "ECMAScript binding" section;
+this is a historical artifact only.
```
Comment (torgo) [inline index.bs:924]: I think this might need clarification. It's a historical artifact only  ... and therefore ...

File: index.bs [hunk 2]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+In particular,
+you can freely rely upon language-specific semantics and conventions,
+with no need to keep things generalized.
+
+<div class="example">
+For example,
+the {{CustomElementRegistry.define()}} method takes a reference to
+a <a href="https://tc39.es/ecma262/#sec-static-semantics-constructormethod">Constructor Method</a>.
+
+This takes advantage of the relatively recent addition of Classes to JavaScript,
```
Comment (kenchris) [inline index.bs:914]: why is classes uppercase?

File: index.bs [hunk 3]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+When designing imperative APIs for the Web,
+use JavaScript.
+In particular,
+you can freely rely upon language-specific semantics and conventions,
+with no need to keep things generalized.
+
+<div class="example">
+For example,
+the {{CustomElementRegistry.define()}} method takes a reference to
+a <a href="https://tc39.es/ecma262/#sec-static-semantics-constructormethod">Constructor Method</a>.
```
Comment (kenchris) [inline index.bs:912]: Is https://tc39.es/ecma262/ the proper link to refer to?

Comment (kenchris) [reply index.bs:912]: I mean like the official one as it refers to a Spanish site (.es)

File: index.bs [hunk 4]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+</div>
 
-This is sometimes a confusing point because [[WEBIDL]] descended from the language-agnostic OMG IDL
-(and at one point, included "Java Bindings"). Even today, the structure of the document contains a
-confusing and redundant division between the "Interface definition language" and the "ECMAScript
-binding". Rest assured that this division is simply a historical artifact of document structure,
-and does not imply anything about the intent of Web IDL in general. The only reason it remains is
-that nobody has taken the time to eradicate it.
+<div class="note">
+JavaScript is also known as [[!ECMASCRIPT]]
```
Comment (kenchris) [inline index.bs:919]: It is standardized under the name ECMAScript

File: index.bs [hunk 5]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
-the output if it has not been garbage collected. This creates behavior which differs based on the
-details of garbage collection, which is strongly discouraged. If {{Document/getElementsByTagName}}
-were designed today, the advice to the specification authors would be to either reliably reuse
-the output, or to produce a new {{HTMLCollection}} each time it's invoked.
-
-The case of {{Document/getElementsByTagName}} is particularly insidious because nothing about
-the API visibly relates to garbage collection details. By contrast, special APIs for this particular
-purpose such as `WeakRef` and `FinalizationGroup` have their GC interaction as a specific part of
-their contract with the developer.
+Don't modify data accessed from JavaScript APIs
```
Comment (kenchris) [inline index.bs:929]: via?

File: index.bs [hunk 6]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+
+A JavaScript Web API is generally a wrapper around
+a feature implemented in a lower-level language,
+such as C++ or Rust.
+Unlike those languages,
+JavaScript guarantees that once a piece of code begins executing,
+it will continue executing until it has completed.
+
+Because of that, JavaScript authors take for granted
+that the data available to a function won’t change
```
Comment (kenchris) [inline index.bs:940]: beneath them?

File: index.bs [hunk 7]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+you can freely rely upon language-specific semantics and conventions,
+with no need to keep things generalized.
+
+<div class="example">
+For example,
+the {{CustomElementRegistry.define()}} method takes a reference to
+a <a href="https://tc39.es/ecma262/#sec-static-semantics-constructormethod">Constructor Method</a>.
+
+This takes advantage of the relatively recent addition of Classes to JavaScript,
+and fact that function pointers are very easy to use in JavaScript.
```
Comment (alice) [inline index.bs:915]:
```suggestion
and fact that function references are very easy to use in JavaScript.
```

File: index.bs [hunk 8]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+
+So if a JavaScript Web API exposes some piece of data,
+such as an object property,
+the user agent must not update that data
+while a JavaScript task is running.
+Instead, if the underlying data changes,
+<a>queue a task</a> to modify the exposed version of the data.
+
+<div class="example">
+If a JavaScript task is has accessed the {{navigator.onLine}} property,
```
Comment (kenchris) [inline index.bs:951]: is has sounds weird

File: index.bs [hunk 9]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+and browser's online status changes,
+the property won't be updated until the next task.
```
Comment (kenchris) [inline index.bs:953]: task is run?

File: index.bs (hunk 7 above)
Comment (cynthia) [inline index.bs]: "and the fact that"?

Also, pointers are a very specific concept; namely one that JS does not have.

File: index.bs [hunk 10]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+in different contexts.
+This means that authors need to write extra code to handle these differences.
+It may also make it more difficult for user agents
+to implement different garbage collection strategies,
+if there is enough code which depends on timing working a particular way.
+
+This means that you shouldn't expose any API that acts as a weak reference,
+e.g. with a property that becomes <code highlight="js">null</code>
+once garbage collection runs.
+Object and data lifetimes in JavaScript code should be predictable.
```
Comment (kenchris) [inline index.bs:974]: JavaScript has WeakRefs today thought

File: index.bs [hunk 11]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
-their contract with the developer.
+Don't modify data accessed from JavaScript APIs
+while a JavaScript <a>event loop</a> is running.
+
+A JavaScript Web API is generally a wrapper around
+a feature implemented in a lower-level language,
+such as C++ or Rust.
+Unlike those languages,
+JavaScript guarantees that once a piece of code begins executing,
+it will continue executing until it has completed.
```
Comment (cynthia) [inline index.bs:937]: Does it really?

File: index.bs (hunk 3 above)
Comment (alice) [reply index.bs:912]: Apparently this is the official one.

File: index.bs (hunk 2 above)
Comment (alice) [reply index.bs:914]: Will fix.

File: index.bs (hunk 10 above)
Comment (kenchris) [reply index.bs:974]: mentioned below

File: index.bs [hunk 12]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+
+So if a JavaScript Web API exposes some piece of data,
+such as an object property,
+the user agent must not update that data
+while a JavaScript task is running.
+Instead, if the underlying data changes,
+<a>queue a task</a> to modify the exposed version of the data.
+
+<div class="example">
+If a JavaScript task is has accessed the {{navigator.onLine}} property,
```
Comment (alice) [inline index.bs:951]:
```suggestion
If a JavaScript task has accessed the {{navigator.onLine}} property,
```

File: index.bs [hunk 13]
```diff
@@ -900,77 +900,102 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+you can freely rely upon language-specific semantics and conventions,
+with no need to keep things generalized.
+
+<div class="example">
+For example,
+the {{CustomElementRegistry.define()}} method takes a reference to
+a <a href="https://tc39.es/ecma262/#sec-static-semantics-constructormethod">Constructor Method</a>.
+
+This takes advantage of the relatively recent addition of Classes to JavaScript,
+and fact that function references are very easy to use in JavaScript.
```
Comment (alice) [inline index.bs:915]:
```suggestion
and the fact that function references are very easy to use in JavaScript.
```

File: index.bs (hunk 11 above)
Comment (alice) [reply index.bs:937]: Are there cases when it wouldn't?

File: index.bs (hunk 6 above)
Comment (alice) [reply index.bs:940]: Hmmm I went with "unexpectedly while the function is running".

File: index.bs (hunk 9 above)
Comment (alice) [reply index.bs:953]: went with "task runs"

File: index.bs (hunk 11 above)
Comment (cynthia) [reply index.bs:938]: The wording sounds like Javascript magically makes the underlying program bug-and-failure-free, which is not true.

JavaScript guarantees that once a piece of code begins executing,

to

The general expectation when using JavaScript is that once a piece of code begins executing,

File: index.bs [hunk 14]
```diff
@@ -900,77 +900,103 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
-purpose such as `WeakRef` and `FinalizationGroup` have their GC interaction as a specific part of
-their contract with the developer.
+Don't modify data accessed from JavaScript APIs
+while a JavaScript <a>event loop</a> is running.
+
+A JavaScript Web API is generally a wrapper around
+a feature implemented in a lower-level language,
+such as C++ or Rust.
+Unlike those languages,
+JavaScript guarantees that once a piece of code begins executing,
```
Comment (alice) [inline index.bs:937]:
```suggestion
Developers can expect when using JavaScript 
that once a piece of code begins executing,
```

File: index.bs (hunk 11 above)
Comment (alice) [reply index.bs:939]: Done with a light reword.

File: index.bs [hunk 15]
```diff
@@ -900,77 +900,104 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
-purpose such as `WeakRef` and `FinalizationGroup` have their GC interaction as a specific part of
-their contract with the developer.
+Don't modify data accessed from JavaScript APIs
+while a JavaScript <a>event loop</a> is running.
+
+A JavaScript Web API is generally a wrapper around
+a feature implemented in a lower-level language,
+such as C++ or Rust.
+Unlike those languages,
+Developers can expect when using JavaScript
```
Comment (alice) [inline index.bs:937]:
```suggestion
developers can expect when using JavaScript 
```

File: index.bs [hunk 16]
```diff
@@ -900,77 +900,104 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+<div class="example">
+If a JavaScript task has accessed the {{navigator.onLine}} property,
+and browser's online status changes,
+the property won't be updated until the next task runs.
+</div>
+
+<h3 id="js-gc">Don't expose garbage collection</h3>
+
+Ensure your JavaScript Web APIs don't behave differently
+depending on the timing of garbage collection.
```
Comment (cynthia) [inline index.bs:961]: I think this statement is semantically different from the opening sentence of the original text.

File: index.bs [hunk 17]
```diff
@@ -900,77 +900,104 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
+
+<div class="example">
+If a JavaScript task has accessed the {{navigator.onLine}} property,
+and browser's online status changes,
+the property won't be updated until the next task runs.
+</div>
+
+<h3 id="js-gc">Don't expose garbage collection</h3>
+
+Ensure your JavaScript Web APIs don't behave differently
```
Comment (alice) [inline index.bs:960]:
```suggestion
Ensure your JavaScript Web APIs don't provide a way
```

File: index.bs (hunk 16 above)
Comment (alice) [inline index.bs:961]:
```suggestion
for an author to deduce the timing of garbage collection.
```

**Review by cynthia — CHANGES_REQUESTED** [review]
One bit needs to be changed, but aside from that LGTM.

File: index.bs (hunk 16 above)
Comment (cynthia) [reply index.bs:961]: s/deduce/know/ ?

File: index.bs [hunk 18]
```diff
@@ -900,77 +900,104 @@ See [[#naming-is-hard]] for general (cross-language) advice on naming.
...
-the output if it has not been garbage collected. This creates behavior which differs based on the
-details of garbage collection, which is strongly discouraged. If {{Document/getElementsByTagName}}
-were designed today, the advice to the specification authors would be to either reliably reuse
-the output, or to produce a new {{HTMLCollection}} each time it's invoked.
-
-The case of {{Document/getElementsByTagName}} is particularly insidious because nothing about
-the API visibly relates to garbage collection details. By contrast, special APIs for this particular
-purpose such as `WeakRef` and `FinalizationGroup` have their GC interaction as a specific part of
-their contract with the developer.
+Don't modify data accessed from JavaScript APIs
```
Comment (alice) [inline index.bs:930]:
```suggestion
Don't modify data accessed via JavaScript APIs
```

**Review by cynthia — APPROVED** [review]

**Review by torgo — APPROVED** [review]
Looks good to me.

**Pull request merged**